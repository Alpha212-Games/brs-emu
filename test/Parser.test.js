const Parser = require("../lib/parser");
const Expr = require("../lib/parser/Expression");
const { Lexeme } = require("../lib/Lexeme");
const OrbsError = require("../lib/Error");

function token(kind, literal) {
    return {
        kind: kind,
        literal: literal,
        line: 1
    };
}

const EOF = token(Lexeme.Eof);

describe("parser", () => {
    afterEach(() => OrbsError.reset());

    describe("primary expressions", () => {
        it("parses literals", () => {
            let numeric = Parser.parse([
                token(Lexeme.Integer, 5),
                EOF
            ]);
            expect(numeric).toEqual(new Expr.Literal(5));

            let parsedString = Parser.parse([
                token(Lexeme.String, "hello"),
                EOF
            ]);
            expect(parsedString).toEqual(new Expr.Literal("hello"));
        });

        it("parses identifiers");

        it("parses expressions in parentheses", () => {
            let withParens = Parser.parse([
                token(Lexeme.Integer, 1),
                token(Lexeme.Plus),
                token(Lexeme.LeftParen),
                token(Lexeme.Integer, 2),
                token(Lexeme.Star),
                token(Lexeme.Integer, 3),
                token(Lexeme.RightParen),
                EOF
            ]);
            expect(withParens).toBeDefined();
            expect(withParens).not.toBeNull();
            expect(withParens).toMatchSnapshot();
        });
    }); // primary expressions

    describe("exponentiation", () => {
        it("parses exponentiation operators", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses repeated exponentiation as left-associative", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 4),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    }); // exponentiation

    describe("unary expressions", () => {
        it("parses unary 'not'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Not),
                token(Lexeme.True),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary 'not'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.True),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses unary '-'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Minus),
                token(Lexeme.Integer, 5),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary '-'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Integer, 5),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    }); // unary expressions
}); // parser