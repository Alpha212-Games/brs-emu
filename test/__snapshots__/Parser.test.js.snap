// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser exponentiation parses exponentiation operators 1`] = `
Binary {
  "left": Literal {
    "value": 2,
  },
  "right": Literal {
    "value": 3,
  },
  "token": Object {
    "kind": 6,
    "line": 1,
    "literal": undefined,
  },
}
`;

exports[`parser exponentiation parses repeated exponentiation as left-associative 1`] = `
Binary {
  "left": Binary {
    "left": Literal {
      "value": 2,
    },
    "right": Literal {
      "value": 3,
    },
    "token": Object {
      "kind": 6,
      "line": 1,
      "literal": undefined,
    },
  },
  "right": Literal {
    "value": 4,
  },
  "token": Object {
    "kind": 6,
    "line": 1,
    "literal": undefined,
  },
}
`;

exports[`parser primary expressions parses expressions in parentheses 1`] = `
Binary {
  "left": Literal {
    "value": 1,
  },
  "right": Grouping {
    "expression": Binary {
      "left": Literal {
        "value": 2,
      },
      "right": Literal {
        "value": 3,
      },
      "token": Object {
        "kind": 9,
        "line": 1,
        "literal": undefined,
      },
    },
  },
  "token": Object {
    "kind": 8,
    "line": 1,
    "literal": undefined,
  },
}
`;

exports[`parser unary expressions parses consecutive unary '-' 1`] = `
Unary {
  "operator": Object {
    "kind": 7,
    "line": 1,
    "literal": undefined,
  },
  "right": Unary {
    "operator": Object {
      "kind": 7,
      "line": 1,
      "literal": undefined,
    },
    "right": Unary {
      "operator": Object {
        "kind": 7,
        "line": 1,
        "literal": undefined,
      },
      "right": Unary {
        "operator": Object {
          "kind": 7,
          "line": 1,
          "literal": undefined,
        },
        "right": Unary {
          "operator": Object {
            "kind": 7,
            "line": 1,
            "literal": undefined,
          },
          "right": Literal {
            "value": 5,
          },
        },
      },
    },
  },
}
`;

exports[`parser unary expressions parses consecutive unary 'not' 1`] = `
Unary {
  "operator": Object {
    "kind": 58,
    "line": 1,
    "literal": undefined,
  },
  "right": Unary {
    "operator": Object {
      "kind": 58,
      "line": 1,
      "literal": undefined,
    },
    "right": Unary {
      "operator": Object {
        "kind": 58,
        "line": 1,
        "literal": undefined,
      },
      "right": Unary {
        "operator": Object {
          "kind": 58,
          "line": 1,
          "literal": undefined,
        },
        "right": Unary {
          "operator": Object {
            "kind": 58,
            "line": 1,
            "literal": undefined,
          },
          "right": Literal {
            "value": true,
          },
        },
      },
    },
  },
}
`;

exports[`parser unary expressions parses unary '-' 1`] = `
Unary {
  "operator": Object {
    "kind": 7,
    "line": 1,
    "literal": undefined,
  },
  "right": Literal {
    "value": 5,
  },
}
`;

exports[`parser unary expressions parses unary 'not' 1`] = `
Unary {
  "operator": Object {
    "kind": 58,
    "line": 1,
    "literal": undefined,
  },
  "right": Literal {
    "value": true,
  },
}
`;
